pkg/repro/repro.go

```diff
@@ -168,46 +243,212 @@ func (ctx *context) reproExtractProg(entries []*prog.LogEntry) (*Result, error)
+		if opts.FaultCall < 0 || opts.FaultCall >= len(ent.P.Calls) {
+			opts.FaultCall = len(ent.P.Calls) - 1
+		}
+		crashed, err := ctx.testProg(ent.P, duration, opts)
+		if err != nil {
+			return nil, err
+		}
+		if crashed {
+			res := &Result{
+				Prog:     ent.P,
+				Duration: duration * 3 / 2,
+				Opts:     opts,
 			}
-			crashed, err := ctx.testProg(ent.P, dur, opts)
+			return res, nil
+		}
+	}
+
+	return nil, nil
+}
+
+func (ctx *context) reproExtractProgBisect(entries []*prog.LogEntry, baseDuration time.Duration) (*Result, error) {
+	Logf(3, "reproducing crash '%v': bisect: bisecting %d programs", ctx.crashDesc, len(entries))
+
+	opts := csource.Options{
+		Threaded:   true,
+		Collide:    true,
+		Repeat:     true,
+		Procs:      ctx.cfg.Procs,
+		Sandbox:    ctx.cfg.Sandbox,
+		EnableTun:  true,
+		UseTmpDir:  true,
+		HandleSegv: true,
+		WaitRepeat: true,
+		Debug:      true,
+		Repro:      true,
+	}
+
+	duration := func(entries int) time.Duration {
+		return baseDuration + time.Duration((entries/4))*time.Second
+	}
+
+	// Check that executing the whole log results in a crash.
+	Logf(3, "reproducing crash '%v': bisect: executing all %d programs", ctx.crashDesc, len(entries))
+	crashed, err := ctx.testProgs(entries, duration(len(entries)), opts)
+	if err != nil {
+		return nil, err
+	}
+	if !crashed {
+		Logf(3, "reproducing crash '%v': bisect: didn't crash", ctx.crashDesc)
+		return nil, nil
+	}
+
+	compose := func(guilty1, guilty2 [][]*prog.LogEntry, chunk1, chunk2 []*prog.LogEntry) []*prog.LogEntry {
+		progs := []*prog.LogEntry{}
+		for _, c := range guilty1 {
+			for _, prog := range c {
+				progs = append(progs, prog)
+			}
+		}
+		progs = append(progs, chunk1...)
+		progs = append(progs, chunk2...)
+		for _, c := range guilty2 {
+			for _, prog := range c {
+				progs = append(progs, prog)
+			}
+		}
+		return progs
+	}
+
+	logGuilty := func(guilty [][]*prog.LogEntry) string {
+		log := "["
+		for i, chunk := range guilty {
+			log += fmt.Sprintf("<%d>", len(chunk))
+			if i != len(guilty)-1 {
+				log += ", "
+			}
+		}
+		log += "]"
+		return log
+	}
+
+	// Bisect the programs to find the ones that cause the crash.
+	guilty := [][]*prog.LogEntry{entries}
+again:
+	Logf(3, "reproducing crash '%v': bisect: guilty chunks: %v", ctx.crashDesc, logGuilty(guilty))
+	for i, chunk := range guilty {
+		if len(chunk) == 1 {
+			continue
+		}
+
+		guilty1 := guilty[:i]
+		guilty2 := guilty[i+1:]
+		Logf(3, "reproducing crash '%v': bisect: guilty chunks split: %v, <%v>, %v", ctx.crashDesc, logGuilty(guilty1), len(chunk), logGuilty(guilty2))
+
+		chunk1 := chunk[0 : len(chunk)/2]
+		chunk2 := chunk[len(chunk)/2 : len(chunk)]
+		Logf(3, "reproducing crash '%v': bisect: chunk split: <%v> => <%v>, <%v>", ctx.crashDesc, len(chunk), len(chunk1), len(chunk2))
+
+		Logf(3, "reproducing crash '%v': bisect: triggering crash without chunk #1", ctx.crashDesc)
+		progs := compose(guilty1, guilty2, []*prog.LogEntry{}, chunk2)
```

Comment:
s/[]*prog.LogEntry{}/nil/

nil is more common and shorter way to create an empty slice.